blockstack/blockstackd.py

```diff
@@ -3529,29 +3586,156 @@ def run_blockstackd():
+        zonefile_dir = os.path.abspath(os.path.join(working_dir, 'zonefiles'))
+        # First, try using a csharp script if dotnet is available (this is the fastest option, ~2 minutes)
+        if which_tools(['dotnet']):
+            cs_script = """
+                using System.Collections.Generic;
+                using System.IO;
+                using System.Linq;
+                class Program {
+                    static void Main(string[] args) {
+                        IEnumerable<string> EnumLines(StreamReader fp) {
+                            while (!fp.EndOfStream) yield return fp.ReadLine();
+                        }
+                        using var reader = new StreamReader(args[0]);
+                        using var writer = new StreamWriter(args[2]);
+                        var entries = EnumLines(reader).Skip(1).AsParallel().Select(line => {
+                            var hash = line.Substring(0, 40);
+                            var zfPath = Path.Join(args[1], hash.Substring(0, 2), hash.Substring(2, 2), hash + ".txt");
+                            return hash + "\\n" + File.ReadAllText(zfPath).Replace("\\n", "\\\\n");
+                        });
+                        foreach (var entry in entries) writer.WriteLine(entry);
+                    }
+                }"""
+            cs_proj = """<Project Sdk="Microsoft.NET.Sdk"><PropertyGroup>
+                    <OutputType>Exe</OutputType><TargetFramework>netcoreapp3.1</TargetFramework>
+                </PropertyGroup></Project>"""
+            tmpdir = tempfile.mkdtemp('blockstack-cs-subdomain-agg')
+            with open(os.path.join(tmpdir, "Program.cs"), "w") as cs_file:
+                cs_file.write(cs_script)
+            with open(os.path.join(tmpdir, "Program.csproj"), "w") as cs_prog_file:
+                cs_prog_file.write(cs_proj)
+            cs_script_args = ["dotnet", "run", "-c=Release", "--", subdomain_csv_path, zonefile_dir, subdomain_zonefile_txt_path]
+            p = subprocess.Popen(cs_script_args, shell=False, cwd=tmpdir, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            p_res = p.communicate()
+            rc = p.returncode
+            if rc != 0:
+                print_status('Subdomain zonefile export via "dotnet" failed with error {}: {}'.format(rc, p_res[1]))
+                sys.exit(1)
+        # Otherwise, try using an awk script, takes ~25 minutes.
+        elif which_tools(['awk']):
+            awk_export_cmd = """awk -F, 'BEGIN {OFS=","} NR>1 {
+                save_rs = RS
+                RS = "^$"
+                file_path = "%s/" substr($1,0,2) "/" substr($1,3,2) "/" $1 ".txt"
+                getline zonefile < file_path
+                gsub(/\\n/,"/n",zonefile)
+                close(file_path)
+                print $1
+                print zonefile
+                RS = save_rs
+            }' %s""" % (zonefile_dir, pipes.quote(subdomain_csv_path))
```

Comment:
Have you confirmed that the awk script and the dotnet script both generate the exact same data?

Reply:
Yes:
```
matt$ dotnet run -c=Release -- /tmp/subdomains.csv /tmp/zonefiles > /tmp/export-dotnet.txt

matt$ time awk -F, 'BEGIN {OFS=","} NR>1 { \
    save_rs = RS; \
    RS = "^$"; \
    file_path = "/tmp/zonefiles/" substr($1,0,2) "/" substr($1,3,2) "/" $1 ".txt";  \
    getline zonefile < file_path; \
    gsub(/\n/,"\\n",zonefile); \
    close(file_path); \
    print $1; \
    print zonefile; \
    RS = save_rs; \
}' /tmp/subdomains.csv > /tmp/export-awk.txt

matt$ sha256sum /tmp/export-dotnet.txt 
35fb3bf4c388ebcc06a9cde7959a4ad1029dc0451307a7e4407783ee146f8887  /tmp/export-dotnet.txt

matt$ sha256sum /tmp/export-awk.txt 
35fb3bf4c388ebcc06a9cde7959a4ad1029dc0451307a7e4407783ee146f8887  /tmp/export-awk.txt
```

Thanks for pointing this out, I found a bug in the awk escaping script, and the dotnet script wasn't outputting in deterministic order. Both now fixed!

Reply:
Good to hear!